File: src/py_mancala/mancala.py

class Mancala:
    def __init__(self):
        print("(python) Mancala::init")

    def get_state(self):
        print("(python) Mancala::get_state")
        return "(python) current state"

    def play_position(self, value):
        print("(python) Mancala::play_position(value={!r})".format(value))

File: src/lib/mancala.cpp

#include "mancala.h"
#include <pybind11/embed.h>

namespace python {
    namespace py = pybind11;

    struct Mancala::Impl {
        py::object mancala;
        py::object get_state;
        py::object play_position;
    };

    Mancala::Mancala() : pImpl(std::make_unique<Impl>()) {
        // State is stored inside the python module
        pImpl->mancala = py::module::import("py_mancala").attr("mancala");
        pImpl->get_state = pImpl->mancala.attr("get_state");
        pImpl->play_position = pImpl->mancala.attr("play_position");
    };
    Mancala::~Mancala() {};

    std::string Mancala::get_state() const {
        return pImpl->get_state().cast<std::string>();
    }

    void Mancala::play_position(const std::string& position) {
        pImpl->play_position(position);
    }

}
